hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOmContainerLocationCache.java

```diff
@@ -0,0 +1,674 @@
+    when(manager.acquireClient(argThat(matchPipeline(DN1))))
+        .thenReturn(mockDn1Protocol);
+    when(manager.acquireClientForReadData(argThat(matchPipeline(DN1))))
+        .thenReturn(mockDn1Protocol);
+
+    when(manager.acquireClient(argThat(matchPipeline(DN2))))
+        .thenReturn(mockDn2Protocol);
+    when(manager.acquireClientForReadData(argThat(matchPipeline(DN2))))
+        .thenReturn(mockDn2Protocol);
+    return manager;
+  }
+
+  private static ArgumentMatcher<Pipeline> matchPipeline(DatanodeDetails dn) {
+    return argument -> argument != null
+        && argument.getNodes().get(0).getUuid().equals(dn.getUuid());
+  }
+
+  private static void createBucket(String volumeName, String bucketName,
+                                   boolean isVersionEnabled)
+      throws IOException {
+    OmBucketInfo bucketInfo = OmBucketInfo.newBuilder()
+        .setVolumeName(volumeName)
+        .setBucketName(bucketName)
+        .setIsVersionEnabled(isVersionEnabled)
+        .build();
+
+    OMRequestTestUtils.addBucketToOM(metadataManager, bucketInfo);
+  }
+
+  private static void createVolume(String volumeName) throws IOException {
+    OmVolumeArgs volumeArgs = OmVolumeArgs.newBuilder()
+        .setVolume(volumeName)
+        .setAdminName("bilbo")
+        .setOwnerName("bilbo")
+        .build();
+    OMRequestTestUtils.addVolumeToOM(metadataManager, volumeArgs);
+  }
+
+  @BeforeEach
+  @SuppressFBWarnings("ST_WRITE_TO_STATIC_FROM_INSTANCE_METHOD")
+  public void beforeEach() {
+    testContainerId++;
+    Mockito.reset(mockScmBlockLocationProtocol, mockScmContainerClient,
+        mockDn1Protocol, mockDn2Protocol);
+    when(mockDn1Protocol.getPipeline()).thenReturn(createPipeline(DN1));
+    when(mockDn2Protocol.getPipeline()).thenReturn(createPipeline(DN2));
+  }
+
+  /**
+   * Verify that in a happy case, container location is cached and reused
+   * in OM.
+   */
+  @Test
+  public void containerCachedInHappyCase() throws Exception {
+    byte[] data = "Test content".getBytes(UTF_8);
+
+    mockScmAllocation(testContainerId, 1L);
+    mockWriteChunkResponse(mockDn1Protocol);
+    mockPutBlockResponse(mockDn1Protocol, testContainerId, 1L, data);
+
+    OzoneBucket bucket = objectStore.getVolume(VOLUME_NAME)
+        .getBucket(BUCKET_NAME);
+
+    // Create keyName1.
+    String keyName1 = "key1";
+    try (OzoneOutputStream os = bucket.createKey(keyName1, data.length)) {
+      IOUtils.write(data, os);
+    }
+
+    mockScmGetContainerPipeline(testContainerId, DN1);
+
+    // Read keyName1.
+    OzoneKeyDetails key1 = bucket.getKey(keyName1);
+    verify(mockScmContainerClient, times(1))
+        .getContainerWithPipelineBatch(newHashSet(testContainerId));
+
+    try (InputStream is = key1.getContent()) {
+      mockGetBlock(mockDn1Protocol, testContainerId, 1L, data, null, null);
+      mockReadChunk(mockDn1Protocol, testContainerId, 1L, data, null, null);
```

Comment:
these two mocks work for both key1 and key2 don't they?

if so, i would move them above the try because they look key1 as is.